### ISSUE TITLE
Stack Overflow analyzer failure after NNBD change

### ISSUE THREAD

**Issue by olesiathoms-wk** [dart-lang/sdk#39937]
After updating to dart 2.7.0 in `built_redux` analysis fails with following exception:
```text
Stack Overflow
#0      Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:602:3)
#1      Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:793:11)
#2      Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#3      Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
#4      Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:747:13)
#5      Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#6      Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
#7      Dart2TypeSystem._isInterfaceSubtypeOf (package:analyzer/src/generated/type_system.dart:1105:14)
#8      Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:814:14)
#9      Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:793:11)
#10     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#11     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
#12     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:747:13)
#13     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#14     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
#15     Dart2TypeSystem._isInterfaceSubtypeOf (package:analyzer/src/generated/type_system.dart:1105:14)
#16     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:814:14)
#17     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:793:11)
#18     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#19     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
#20     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:747:13)
#21     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#22     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
#23     Dart2TypeSystem._isInterfaceSubtypeOf (package:analyzer/src/generated/type_system.dart:1105:14)
#24     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:814:14)
#25     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:793:11)
#26     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#27     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
#28     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:747:13)
#29     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#30     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
#31     Dart2TypeSystem._isInterfaceSubtypeOf (package:analyzer/src/generated/type_system.dart:1105:14)
#32     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:814:14)
#33     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:793:11)
#34     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#35     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
#36     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:747:13)
#37     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#38     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
#39     Dart2TypeSystem._isInterfaceSubtypeOf (package:analyzer/src/generated/type_system.dart:1105:14)
#40     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:814:14)
#41     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:793:11)
#42     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#43     Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
...
...
#6602   Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:715:14)
#6603   Dart2TypeSystem.isSubtypeOf (package:analyzer/src/generated/type_system.dart:708:14)
#6604   _OverrideChecker._findCovariantChecksForMember.<anonymous closure> (package:analyzer/src/task/strong/checker.dart:1614:20)
#6605   FunctionTypeImpl.relateTypeFormals2 (package:analyzer/src/dart/element/type.dart:875:20)
#6606   _OverrideChecker._findCovariantChecksForMember (package:analyzer/src/task/strong/checker.dart:1610:34)
#6607   _OverrideChecker._findCovariantChecks (package:analyzer/src/task/strong/checker.dart:1542:9)
#6608   _OverrideChecker._checkForCovariantGenerics (package:analyzer/src/task/strong/checker.dart:1488:18)
#6609   _OverrideChecker.check (package:analyzer/src/task/strong/checker.dart:1424:5)
#6610   CodeChecker.visitClassDeclaration (package:analyzer/src/task/strong/checker.dart:330:22)
#6611   ClassDeclarationImpl.accept (package:analyzer/src/dart/ast/ast.dart:1512:49)
#6612   NodeListImpl.accept (package:analyzer/src/dart/ast/ast.dart:7516:20)
#6613   CompilationUnitImpl.visitChildren (package:analyzer/src/dart/ast/ast.dart:2111:21)
#6614   CodeChecker.visitCompilationUnit (package:analyzer/src/task/strong/checker.dart:351:10)
#6615   LibraryAnalyzer._computeVerifyErrors (package:analyzer/src/dart/analysis/library_analyzer.dart:368:13)
#6616   LibraryAnalyzer.analyzeSync.<anonymous closure>.<anonymous closure> (package:analyzer/src/dart/analysis/library_analyzer.dart:165:9)
#6617   _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:377:8)
#6618   LibraryAnalyzer.analyzeSync.<anonymous closure> (package:analyzer/src/dart/analysis/library_analyzer.dart:164:13)
#6619   _PerformanceTagImpl.makeCurrentWhile (package:analyzer/src/generated/utilities_general.dart:258:15)
#6620   LibraryAnalyzer.analyzeSync (package:analyzer/src/dart/analysis/library_analyzer.dart:163:34)
#6621   LibraryAnalyzer.analyze.<anonymous closure> (package:analyzer/src/dart/analysis/library_analyzer.dart:114:14)
#6622   _PerformanceTagImpl.makeCurrentWhile (package:analyzer/src/generated/utilities_general.dart:258:15)
#6623   LibraryAnalyzer.analyze (package:analyzer/src/dart/analysis/library_analyzer.dart:113:43)
#6624   AnalysisDriver._computeAnalysisResult.<anonymous closure> (package:analyzer/src/dart/analysis/driver.dart:1261:63)
#6625   PerformanceLog.run (package:analyzer/src/dart/analysis/performance_logger.dart:34:15)
#6626   AnalysisDriver._computeAnalysisResult (package:analyzer/src/dart/analysis/driver.dart:1237:20)
#6627   AnalysisDriver.performWork (package:analyzer/src/dart/analysis/driver.dart:1047:17)
#6628   _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:71:64)
#6629   _rootRunUnary (dart:async/zone.dart:1132:38)
#6630   _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#6631   _FutureListener.handleValue (dart:async/future_impl.dart:137:18)
#6632   Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:678:45)
#6633   Future._propagateToListeners (dart:async/future_impl.dart:707:32)
#6634   Future._addListener.<anonymous closure> (dart:async/future_impl.dart:387:9)
#6635   _rootRun (dart:async/zone.dart:1124:13)
#6636   _CustomZone.runGuarded (dart:async/zone.dart:1021:19)
#6637   _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
#6638   _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#6639   _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#6640   _Timer._runTimers (dart:isolate-patch/timer_impl.dart:391:30)
#6641   _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart)
```
Dart 2.6.1 doesn't have this problem.
Going through changes made between 2.6.1. and 2.7.0. I found out that switch to NNBD subtyping rules caused analyzer to fail https://github.com/dart-lang/sdk/commit/70b1cedd87191d11293b5d1192fada7fdbc8fc04 

I think `combinedNested` functions causing this problem from `built_redux` side:
https://github.com/davidmarne/built_redux/blob/master/lib/src/middleware.dart#L52-L61
https://github.com/davidmarne/built_redux/blob/master/lib/src/reducer_builder.dart#L31-L36
 specifically extensions on `Built` and `Builder`.
Was this expected? if so, do we need to update our code to follow new rules?

```
~/built_redux (master ✔) ᐅ dartanalyzer  --version
dartanalyzer version 2.7.0
~/built_redux (master ✔) ᐅ dart --version
Dart VM version: 2.7.0 (Fri Dec 6 16:26:51 2019 +0100) on "macos_x64"
```

**Comment by bwilkerson** [comment]
@scheglov

**Comment by scheglov** [comment]
No, this is not expected, any code should be handled without crashes in analyzer.

Unfortunately I cannot reproduce the issue.
I cloned the repo `https://github.com/davidmarne/built_redux`.

Then I run `flutter analyzer` while in `/Users/scheglov/dart/built_redux`:
```
Upgrading Flutter from /Users/scheglov/Source/flutter...
Updating cf37c2cd0..27321ebba
 24 files changed, 651 insertions(+), 218 deletions(-)
Flutter is already up to date on channel stable
Flutter 1.12.13+hotfix.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 27321ebbad (2 weeks ago) • 2019-12-10 18:15:01 -0800
Engine • revision 2994f7e1e6
Tools • Dart 2.7.0
scheglov@scheglov-macbookpro2:~/dart/built_redux (master)$ flutter analyze
Downloading package sky_engine...                                   0.4s
Downloading common tools...                                         1.2s
Downloading common tools...                                         1.3s
Downloading darwin-x64 tools...                                     3.5s
Running "flutter pub get" in built_redux...                         0.6s
Analyzing built_redux...
No issues found! (ran in 10.2s)
```

I also tried the bleeding edge of Dart SDK, it still can be analyzer without the Stack Overflow.

It also does not crash with the `master` channel of Flutter, which is `Dart 2.8.0` AFAIK.

Does the way I tried to reproduce looks right to you? Do I need a different version of `built_redux`?

**Comment by olesiathoms-wk** [comment]
you should be able to reproduce it on master.
 Is there a difference between `flutter analyze` and `dartanalyzer`?
 I just tried using flutter and it worked, but at the same time `dartanalyzer` fails.
```
~/built_redux (master ✔) ᐅ flutter analyze
Analyzing built_redux...
No issues found! (ran in 9.4s)
```

**Comment by scheglov** [comment]
`flutter analyze` runs Dart Analysis Server, and `dartanalyzer` is a separate CLI utility. In principle there should be no difference, both use `analyzer` for actual analysis, just different ways to build analysis contexts. But they might be have different versions - `dartanalyzer` is probably from normal, non-Flutter Dart SDK. Please check the Dart SDK version. FWIW, this works for me:

```
scheglov@scheglov-macbookpro2:~/dart/built_redux (master)$ dartanalyzer --version
dartanalyzer version 2.8.0-edge.3ab95a114c6ce260f1e54752b4ebc8c322173a8a
scheglov@scheglov-macbookpro2:~/dart/built_redux (master)$ dartanalyzer .
Analyzing built_redux...
No issues found!
```

**Comment by olesiathoms-wk** [comment]
I was on `2.7.0` and checked `2.8.0-dev.0.0` as well, they both were failing. I just checked the same version that you are on and it's working. I'll do a little more exploring, but I think it was fixed in some commit over last two weeks

**Comment by olesiathoms-wk** [comment]
I went through commits made after `2.8.0-dev.0.0` and 
```
Dart VM version: 2.8.0-edge.b3c2eb9f698408bca8fa35ca5cc6c6f088098ec1 (Thu Dec 12 23:54:41 2019 +0000) on "macos_x64"
``` 
 introduced fix for the first time. Is it possible to release it as a patch for 2.7?

**Comment by jcollins-g** [comment]
Verified that b3c2eb9f698408bca8fa35ca5cc6c6f088098ec1 does contain the fix, but it does not patch cleanly to 2.7.0.  The conflict looks fairly trivial and a naive conflict resolution results in an analyzer that can at least analyze built_redux cleanly.

**Comment by aaronlademann-wf** [comment]
@bwilkerson @scheglov @jcollins-g any chance this gets patched into 2.7.x since _"the conflict looks fairly trivial and a naive conflict resolution results in an analyzer that can at least analyze built_redux cleanly."_?

Or is the 2.8.0 release imminent? 

This causes `built_redux`, and all packages that depend on it - to not produce API documentation when `dartdoc` is run using SDK version `2.7.x` w/ analyzer version `0.39.4`.

cc/ @greglittlefield-wf @md-weber @kevmoo @robbecker-wf

**Comment by kevmoo** [comment]
Is this an SDK issue or a `pkg:analyzer` issue?

**Comment by jcollins-g** [comment]
@kevmoo Yes to both.  The report concerns the version of analyzer shipped as part of the 2.7 SDK.

**Comment by kevmoo** [comment]
CC @franklinyow @athomas 

We could consider a 2.7.2 release for this.

@aaronlademann-wf – have ya'll looked at rolling to 2.8.0-dev.1.0 as an alternative?

**Comment by aaronlademann-wf** [comment]
@kevmoo I can do that locally - but that won't fix the underlying issue - which is that when a package that depends on `built_redux` is published to `pub.dev`, the generation of the API documentation will fail.

**Comment by kevmoo** [comment]
> @kevmoo I can do that locally - but that won't fix the underlying issue - which is that when a package that depends on `built_redux` is published to `pub.dev`, the generation of the API documentation will fail.

Good to know. I'll work with @franklinyow on this.

**Comment by kevmoo** [comment]
Fixed in https://github.com/dart-lang/sdk/commit/b3c2eb9f698408bca8fa35ca5cc6c6f088098ec1

Will be available in the first Dart 2.8 stable release (and many recent 2.8.0-dev releases)
